### ISSUE TITLE
Add a proper test suite

### ISSUE THREAD

**Issue by treeowl** [haskell/primitive#89]
I just went to try to add QuickCheck or SmallCheck properties to the test suite, and discovered that there isn't a proper test suite. Ack! No wonder the `Foldable` instance for `Array` managed to be wrong for so long! I'll be happy to write some properties, but I don't have the time right now to set up the framework.

**Comment by RyanGlScott** [comment]
Alas, this is not as easy to accomplish as it sounds. `QuickCheck` in particular depends on `tf-random`, which in turn depends on `primitive` itself, so you quickly run into https://github.com/haskell/cabal/issues/5200 if you try to make `primitive`'s test suite depend on it.

That being said, there is a relatively serviceable workaround: turn the test suite into its own `.cabal` file and make it its own project, distinct from `primitive` itself. That way, build tools can first build `primitive` and then build `tf-random` _et al._ using the locally built `primitive`, avoiding any dependency cycles. I've used this hack before to make a [test suite for `transformers-compat`](https://github.com/ekmett/transformers-compat/pull/33), and it seems to work okay.

**Comment by RyanGlScott** [comment]
I've opened #90 to implement the suggestion in https://github.com/haskell/primitive/issues/89#issuecomment-374210328. Once that is merged (pending Travis giving a green check mark), that should make setting up `QuickCheck` tests far, far easier.

**Comment by treeowl** [comment]
Great! Have you had a chance to look at my folds and traversals?

On Mar 19, 2018 10:02 AM, "Ryan Scott" <notifications@github.com> wrote:

I've opened #90 <https://github.com/haskell/primitive/pull/90> to implement
the suggestion in #89 (comment)
<https://github.com/haskell/primitive/issues/89#issuecomment-374210328>.
Once that is merged (pending Travis giving a green check mark), that should
make setting up QuickCheck tests far, far easier.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/haskell/primitive/issues/89#issuecomment-374222243>,
or mute
the thread

**Comment by andrewthad** [comment]
@RyanGlScott Thanks for moving the test suite into its own project. I'm going to work on adding some property tests.

**Comment by treeowl** [comment]
Andrew, please let me know if/when you start, and with what. I'll do the
same.

**Comment by andrewthad** [comment]
I'm working on it now. I'm going to use `tasty` + `quickcheck` + `quickcheck-classes` just as an easy way to get access to some of the typeclass property tests we're interested in.

**Comment by RyanGlScott** [comment]
Thanks @andrewthad!

Do be aware that `primitive` has a support window back to GHC 7.4, so using `quickcheck-classes` in its current state would shorten that window somewhat (as it only supports GHC 7.8 and later at the moment).

**Comment by treeowl** [comment]
Sounds great! I'll be happy to help expand it once you get it going.

**Comment by andrewthad** [comment]
@RyanGlScott Good point. Let me see if I can get quickcheck-classes to support older GHCs. I don't think there's any fundamental reason that it can't.

**Comment by andrewthad** [comment]
It was more difficult than I expected, but I now have `quickcheck-classes` building with all GHCs back to 7.4. I'll get the tests together now.

**Comment by treeowl** [comment]
Huzzah! Unfortunately, `quickcheck-classes` is not yet really up to the job. It's missing `Traversable` altogether (which I have a PR to reimplement). It's also missing the `foldl1` and `foldr1` methods; we had some trouble with `foldl1`! I also don't really understand its tests. I would think the `Monad` laws, for instance, would be tested using `Test.QuickCheck.Function`, but they're not. Why not something roughly like this:

```haskell
monadAssoc ::  ... => m A -> Fun A (m B) -> Fun B (m C) -> Property
monadAssoc m f g = (m >>= \x -> apply f x >>= apply g) `equivalent`
                         ((m >>= apply f) >>= apply g)

For `Traversable`, I suggest you consider the answers to [my long-ago SO question on the topic](https://stackoverflow.com/questions/34932522/how-can-i-test-functions-polymorphic-over-applicatives).

**Comment by andrewthad** [comment]
I've got a start on it here: https://github.com/haskell/primitive/pull/99. Already, we can see that it fails the laws for `Functor`, which is expected.

@treeowl I agree that it's currently missing several things. I had never done `Traverable` since those are normally derived. I had only done `Foldable` (also normally easy to derive) as an exercise since the strict variants of the fold functions are tricky to test correctly. I've never used `Test.QuickCheck.Function`, and it's almost certainly a better way to test some of the monad properties than what I came up with.

**Comment by cartazio** [comment]
I’m nearly done with getting random updated in shape.  But I could pull out
split mix so we have a copy in primitive too. That would address circular
deps perhaps ?

**Comment by treeowl** [comment]
For `Traversable`, you need to test the identity law, which should be easy, and the composition law, which will be harder. Also, like `checkers`, you'll want to verify that `foldMapDefault` and `fmapDefault` match `foldMap` and `fmap`. If you're adding detailed strictness checks as well (see `ChasingBottoms`), you'll probably want to check the naturality law too, but that sort of detail can surely wait.

**Comment by treeowl** [comment]
@cartazio, I don't see how that could help. `QuickCheck` has loads of dependencies. We could try a different test framework, I suppose.

**Comment by treeowl** [comment]
@andrewthad, are we ready to close this, or are there still non-trivial functions that lack adequate tests?

**Comment by andrewthad** [comment]
We still aren't testing the following typeclass instances: `MonadZip`, `Alternative`, `MonadPlus`, and `MonadFix`. Of those, only `MonadFix` has what I would describe as a trivial implementation. All the others are at least as complicated as `fmap`. However, I would prefer closing out this issue, since the objective "Add a proper test suite" is complete, and opening a new one instead to track progress on those.

**Comment by treeowl** [comment]
Done! I opened up an issue to track test suite expansion.
